User/rschloen/command blocks

rschloen opened the pull request:
**Add actuator command blocks with unit test**

Added `buildCommandBlock` function to generate the entire command to be sent to reader during mouse event on actuator button. Takes in id of active actuator and uses bit math to convert to the binary representation and set the actuator command blocks (block0_31, block32_63,...). The function can handle up to 128 actuator blocks, but doesn't handle multiple active actuators.

A test was created to check the bit math in `buildCommandBlock`. Dependencies and configurations were added to package.json, babel.config.js, jest.config.js to setup test environment. Tests are in test folder under src/. Manual.test.js has one test that loops the number of actuators to test and compares the commands generated by `buildCommandBlock` to hard coded ground truth commands that are known to be correct. During the test, `buildCommandBlock` is called by buttons for each actuator generated only for the test, and the commands are in a <p> tag generated only for the test (see lines 466-471 for Manual.svelte). Ground truth commands are available for 128 actuators (only first 36 used at the moment). 

Started on timing blocks as well. Timing blocks are handled with reactive statements as the bit math is a little simpler. Math is carried over from the React app and does not have testing yet.

rschloen committed:
Added function generate actuator command blks

Takes activeHexagon value and uses bitshifting to convert to binary
representation. Only for one active actuator.

Added reactive variables for actuator blocks and full
command.

rschloen committed:
Added timing blocks to command blk generator fn

The bitshifting to generate the timing commands from slider values was
added to buildCommandBlocks function.

rschloen committed:
Made timing block reactive

Have timing block be reactive might make updating block easier with both
in specifying mode and infered mode for timing.

rschloen committed:
Added unit testing for actuator block commands

Added buttons and a paragraph to html that call buildCommandBlock with
each hexagon's id to test bit math in the function. Buttons and text are
generated only when the test is being run.

Created test file with a test that loops through each actuator and
compares the hard coded ground truth command to the command generated
from the bit math in the buildCommandBlock function. Ground truth
available for all 128 actuators. Commands are binary strings.

Add babel and jest configs to setup test environment.

To run test: npm run test

trueb2 commented on package.json:
```diff
@@ -1,19 +1,41 @@
   "version": "1.0.0",
+  "jest": {
+    "transform": {
+      "^.+\\.svelte$": "svelte-jester"
```
If jest is configured for svelte, can we write svelte tests rather than js tests?

trueb2 replied:
I think the unit test is confirmation of what I was originally questioning here since the unit test actually does a svelte render and verifies attributes in js.

trueb2 commented on package.json:
```diff
@@ -1,19 +1,41 @@
-    "start": "sirv public --single"
+    "start": "sirv public --single",
+    "test": "jest src",
+    "test:watch": "npm run test -- --watch"
```
Did you have it auto running tests every time you saved?

rschloen replied:
I wasn't initially but I have tested it and it works to auto run it. I just had to increase the number of watchers.

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -1,7 +1,8 @@
 <script>
-import { Router, Link, Route } from "svelte-routing";
+import  { Router, Link, Route } from "svelte-routing";
```
```suggestion
import { Router, Link, Route } from "svelte-routing";
```

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -104,14 +105,43 @@ $: lfDutyCycle = 50;
+$: single_pulse_block = [0,0,0,0];
+$: hfOn = hfPeriod * (hfDutyCycle/100);
+$: hf_block = [hfOn & 0x000000ff, (hfOn & 0x0000ff00) >> 8, hfPeriod & 0x000000ff,(hfPeriod & 0x0000ff00) >> 8];
+$: lfOn = lfPeriod * (lfDutyCycle/100);
+$: lf_block = [lfOn & 0x000000ff, (lfOn & 0x0000ff00) >> 8, lfPeriod & 0x000000ff, (lfPeriod & 0x0000ff00) >> 8];
```
While integrating with the feig reader currently, these blocks are set to 0xff. I think this has something to do with a firmware bug. Since everything works for the actuators without the low frequency settings, this hasn't been a priority fix.

It would be good to add a button to show and use low frequency configs, or hide and assume 0xff. Intensity calculated on [0,100] scale would set low frequency to 0xff, high frequency to something small like 10ms and only modulate the high frequency duty cycle.

rschloen replied:
I added the button(Link) for specifying low frequency per your suggestion. Also added a function that is called when the page is loaded after a link is clicked and fixes the appropriate periods/duty cycles.

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -104,14 +105,43 @@ $: lfDutyCycle = 50;
+
+//Leaving this reactive 
+$: act_command = `{ "ActuatorsCommand": {
+    "fabric_name": "${devices[activeDevice]}",
```
There was a small bug in the React version where device/fabric name would have white space removed, effectively changing the name we submit. It would reduce network traffic very marginally but caused a mismatch on whitespace. 

When configuring device names, I would recommend trimming leading and trailing spaces and replace the remaining white space with underscores before creating the command to send to the CLI.

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -211,6 +241,62 @@ $: drawableHexagons = hexagons.map(({id, row, col}) => {
     return {id, x, y, width, height, points: points.join(' '), color};
 });
 
+function buildCommandBlocks(active){
```
```suggestion
function buildCommandBlocks(active) {
```

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -211,6 +241,62 @@ $: drawableHexagons = hexagons.map(({id, row, col}) => {
 
+function buildCommandBlocks(active){
+    let actuator = active;
+    console.log(actuator);
```
```suggestion
```

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -211,6 +241,62 @@ $: drawableHexagons = hexagons.map(({id, row, col}) => {
+        block32_63 = [0,0,0,0];
+        block64_95 = [0,0,0,0];
+        block96_127 = [0,0,0,0];
+        // block0_31[3] = parseInt(binActuator.slice(0,8),2); //decimal representation of binary command
```
```suggestion
```

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -211,6 +241,62 @@ $: drawableHexagons = hexagons.map(({id, row, col}) => {
+        block96_127 = [0,0,0,0];
+    } else if (actuator >= 64 && actuator < 95){
+        shiftActuator = Math.abs(1 << (actuator - 64)); 
+        binActuator = shiftActuator.toString(2).padStart(32,'0');
```
It would be nice if there was a short helper function that is tested. We may want to use such a utility for grabbing bytes elsewhere. It seems like with the shifting that you did, we may be able to have it assign all the blocks. without the if statements.

```
function getBytesForActuator(actuator) {
   const shiftActuator = ...;
   const binActuator = ...;
   return [...Array(8).keys()].map(i => binActuator.slice(i * 8, (i+1) * 8)).reversed();
}

const [b0, b1, b2, b3, ...rest] = getNumericActuator(actuator);
block0_31 = [b0, b1, b2, b3]

```

rschloen replied:
I added the helper function, and agree that it may be helpful in the future. At the moment it does still use the if statements, but your method of slicing into an array did help clean it up at least. I watched the actuator command test while working on the helper function so I confirmed the helper function works as before.

trueb2 commented on src/routes/Manual.svelte:
```diff
@@ -368,14 +455,20 @@ function handleMouseUp(event) { handleTouchEnd(event); }
                           stroke={hexagon.color}
-                          strokeWidth="5px"
+                          stroke-width="5px"
                           strokeLinejoin="miter"
```
The casing is different for React so this line is probably ignored as well.

rschloen replied:
Changing stroke-width does change the width of the hexagons' outline.

trueb2 commented on src/test/Manual.test.js:
```diff
@@ -0,0 +1,148 @@
+];
+
+
+test('Sends correct Actuator command',async () => {
+    const { getByText } = render(Manual,{test_ui: true})
+    // Set output to watch
+    const output = getByText('Test: 0,0,0,0, 0,0,0,0, 0,0,0,0, 0,0,0,0')
+    //Loop through all actuators
+    for (let index = 0; index < 36; index++) {
+        const button = getByText('Button ' + index.toString())
+        await fireEvent.click(button)
+        expect(output).toHaveTextContent(groundTruth[index])
+    }
+
+})
```
Nice. The passing of props can be very extendable from here. If we start really building up a lot of tests, this pattern will create problems with our deployment bundle size. We have a loooong way to go before we might hit that though.

trueb2 reviewed (changes requested):
Looks good 👍 Take a quick pass to clean up a couple things and ping me to merge. If you have questions, we can merge tomorrow morning.

I would recommend changing the PR names to be a super short description of the changes going in. If you open a PR to start looking at a diff or have conversations, you can put a 'WIP: the rest of the title goes here' title name on the PR to note that it is still a work in progress.

rschloen committed:
Fixed style errors and cleaned up unneeded code

rschloen committed:
Added helper function to do actuator bit shifting

Replaces if statements found in buildCommandBlock and returns array of
actuator block commands. Uses more efficient slicing of 128 char string.

rschloen committed:
Groundtruth updated to match new output

the format if the command block's output was changed with addition of
getBytesForActuators.

rschloen committed:
Added additional controls to set timing blocks

Infer Timing config Link now fixes lfPeriod, lfDutyCycle, hfPeriod to
reasonable values and uses hfDutyCycle for intensity.

Specify Timing config Link now has additional button(link) to specify
low freq(lf) timing. When lf timing is inferred, lf period and DC are
fixed.

Specify lf link gives full control of low and high freq timing.

Control for Single Pulse duration was added, but bit math still needed.

setTimingBlock function called when link is changed and fixes the
appropriate values for infered, and lf inferred
